src/zui/ZUIEditor/README.md

```diff
@@ -0,0 +1,150 @@
+
+### Communicating from UI to extension
+
+UI components are used to present a graphical interface through which the user
+can interact with the editor, so they need to be able to trigger changes in the
+editor by sending messages to it.
+
+Components send messages to extensions using commands, that are exposed by the
+editor using the `@command` decorator in the extension.
+
+Example: The `ImageExtensionUI` presents the file picker and updates the image
+block by invoking the `setImageFile()` command.
+
+```tsx
+const { setImageFile } = useCommands();
+
+<FileLibraryDialog
+  onSelectFile={(file) => {
+    setImageFile(file);
+  }}
+  // ...
+/>;
+```
+
+The `setImageFile()` command is exposed by the `ImageExtension` extension class,
+using Remirrors mechanism for creating ProseMirror commands. Read more about it
+[in their tutorial](https://www.remirror.io/docs/getting-started/custom-extension#add-commands).
+
+```ts
+  @command()
+  setImageFile(file: ZetkinFile | null): CommandFunction {
+    return (props) => {
+      props.dispatch?.(
+        props.tr.setNodeAttribute(
+          props.state.selection.$from.pos,
+          'src',
+          file?.url ?? null
+        )
+      );
+      return true;
+    };
+  }
+```
+
+### Communicating from extension to UI
+
+For UI components to be able to present accurate UI they need to know the state
+of the underlying extension. Extensions can notify the UI of changes using
+events.
+
+Events are defined as part of the extension options using the `Handler` type,
+and then invoked by the extension via `this.options`.
+
+Example: `ImageExtensionUI` listens to an `onCreate` event that is dispatched by
+the extension when it creates a new image block.
+
+```ts
+type ImageOptions = {
+  onCreate?: Handler<() => void>; // <-- Define the event type
+};
+
+@extension({
+  customHandlerKeys: [],
+  defaultOptions: {},
+  handlerKeys: ['onCreate'], // <-- Enable the event
+  staticKeys: [],
+})
+export default class ImageExtension extends NodeExtension<ImageOptions> {
+  createAndPick() {
+    const node = this.type.create();
+    this.options.onCreate(); // <-- Invoke the event
+    return node;
+  }
+  // ...
+}
+```
+
+The component listens for the event using `useExtensionEvent()`:
+
+```ts
+useExtensionEvent(ImageExtension, 'onCreate', () => {
+  setFileDialogOpen(true);
+});
+```
+
+### Reading editor state in components
+
+UI components use the `Remirror` hooks directly to read editor state, such as
+the current position of the cursor and selection, what blocks exist in the
+document, etc.
+
+## Interface
+
+The `ZUIEditor` component exposes these properties:
+
+- `enableX` (where X is a feature) configures the editor to include
+  functionality, e.g. `enableImage` to enable the image block
+- `onChange` is called whenever content changes and passes the content in the
+  Zetkin block format
+- `initialContent` is the initial content defined using the Zetkin block format
```

Comment:
These have not been implemented yet. We can add them when we take on serialization/deserialization of content.